src/shared/kafka-producer-wrapper.ts

```diff
@@ -0,0 +1,90 @@
+import { StatsD } from 'hot-shots'
+import { Producer, ProducerRecord } from 'kafkajs'
+
+import { PluginsServerConfig } from '../types'
+import { timeoutGuard } from './ingestion/utils'
+import { instrumentQuery } from './metrics'
+
+export class KafkaProducerWrapper {
+    /** Kafka producer used for syncing Postgres and ClickHouse person data. */
+    producer: Producer
+    /** StatsD instance used to do instrumentation */
+    statsd: StatsD | undefined
+
+    lastFlushTime: number
+    currentBatch: Array<ProducerRecord>
+    currentBatchSize: number
+
+    flushFrequencyMs: number
+    maxQueueSize: number
+    maxBatchSize: number
+
+    flushInterval: NodeJS.Timeout
+
+    constructor(producer: Producer, statsd: StatsD | undefined, serverConfig: PluginsServerConfig) {
+        this.producer = producer
+        this.statsd = statsd
+
+        this.lastFlushTime = Date.now()
+        this.currentBatch = []
+        this.currentBatchSize = 0
+
+        this.flushFrequencyMs = serverConfig.KAFKA_FLUSH_FREQUENCY_MS
+        this.maxQueueSize = serverConfig.KAFKA_PRODUCER_MAX_QUEUE_SIZE
+        this.maxBatchSize = serverConfig.KAFKA_MAX_MESSAGE_BATCH_SIZE
+
+        this.flushInterval = setInterval(() => this.flush(), this.flushFrequencyMs)
+    }
+
+    async queueMessage(kafkaMessage: ProducerRecord): Promise<void> {
+        const messageSize = this.getMessageSize(kafkaMessage)
+
+        if (this.currentBatchSize + messageSize > this.maxBatchSize) {
+            await this.flush(kafkaMessage)
+        } else {
```

Comment:
I was a bit confused by these lines at first. It seems that the `kafkaMessage` is appended to the flushed queue... whereas in reality it's appended to a new and empty queue.

I think the code would read better if the flush takes no arguments and the `else` part of this will run always?

Reply:
It's hard to implement it that way without introducing any races though. 

There were three cases we want to flush:
1. Message size + current batch exceeds max size
2. Too much time passed
3. Too many messages queued.

Cases 2-3 we handle flushing by adding the message to queue, resetting the queue before sending it on the way.

1 is slightly tricky - we want to flush messages that were queued _before_ this message and set the queue to be this single message. A naive solution would do something like 

```ts
await this.flush()
this.queue.push(kafkaMessage)
```

But this introduces a new race condition that could result in too large messages getting sent.

So the parameter was added to combat it. It's slightly smelly but has minimal leakage and is tested so  ¯\_(ツ)_/¯

Reply:
Ok. Probably would be good to add some deodrant comment to explain this

Reply:
Done